src/sunadaptcontroller/soderlind/sunadaptcontroller_soderlind.c

```diff
@@ -0,0 +1,393 @@
+  }
+
+  /* return with success */
+  return SUNADAPTCONTROLLER_SUCCESS;
+}
+
+int SUNAdaptController_Reset_Soderlind(SUNAdaptController C)
+{
+  SODERLIND_EP(C)  = RCONST(1.0);
+  SODERLIND_EPP(C) = RCONST(1.0);
+  SODERLIND_HP(C)  = RCONST(1.0);
+  SODERLIND_HPP(C) = RCONST(1.0);
+  SODERLIND_FIRSTSTEPS(C) = 0;
+  return SUNADAPTCONTROLLER_SUCCESS;
+}
+
+int SUNAdaptController_SetDefaults_Soderlind(SUNAdaptController C)
+{
+  SODERLIND_K1(C)   = DEFAULT_K1;
+  SODERLIND_K2(C)   = DEFAULT_K2;
+  SODERLIND_K3(C)   = DEFAULT_K3;
+  SODERLIND_K4(C)   = DEFAULT_K4;
+  SODERLIND_K5(C)   = DEFAULT_K5;
+  SODERLIND_BIAS(C) = DEFAULT_BIAS;
+  return SUNADAPTCONTROLLER_SUCCESS;
+}
+
+int SUNAdaptController_Write_Soderlind(SUNAdaptController C, FILE *fptr)
+{
+  fprintf(fptr, "Soderlind SUNAdaptController module:\n");
+#if defined(SUNDIALS_EXTENDED_PRECISION)
+  fprintf(fptr, "  k1 = %32Lg\n", SODERLIND_K1(C));
+  fprintf(fptr, "  k2 = %32Lg\n", SODERLIND_K2(C));
+  fprintf(fptr, "  k3 = %32Lg\n", SODERLIND_K3(C));
+  fprintf(fptr, "  k4 = %32Lg\n", SODERLIND_K4(C));
+  fprintf(fptr, "  k5 = %32Lg\n", SODERLIND_K5(C));
+  fprintf(fptr, "  bias factor = %22Lg\n", SODERLIND_BIAS(C));
+  fprintf(fptr, "  previous error = %22Lg\n", SODERLIND_EP(C));
+  fprintf(fptr, "  previous-previous error = %22Lg\n", SODERLIND_EPP(C));
+  fprintf(fptr, "  previous step = %22Lg\n", SODERLIND_HP(C));
+  fprintf(fptr, "  previous-previous step = %22Lg\n", SODERLIND_HPP(C));
+#else
+  fprintf(fptr, "  k1 = %16g\n", SODERLIND_K1(C));
+  fprintf(fptr, "  k2 = %16g\n", SODERLIND_K2(C));
+  fprintf(fptr, "  k3 = %16g\n", SODERLIND_K3(C));
+  fprintf(fptr, "  k4 = %16g\n", SODERLIND_K4(C));
+  fprintf(fptr, "  k5 = %16g\n", SODERLIND_K5(C));
+  fprintf(fptr, "  bias factor = %16g\n", SODERLIND_BIAS(C));
+  fprintf(fptr, "  previous error = %16g\n", SODERLIND_EP(C));
+  fprintf(fptr, "  previous-previous error = %16g\n", SODERLIND_EPP(C));
+  fprintf(fptr, "  previous step = %16g\n", SODERLIND_HP(C));
+  fprintf(fptr, "  previous-previous step = %16g\n", SODERLIND_HPP(C));
+#endif
+  fprintf(fptr, "  firststeps = %i\n", SODERLIND_FIRSTSTEPS(C));
+  return SUNADAPTCONTROLLER_SUCCESS;
+}
+
+int SUNAdaptController_SetErrorBias_Soderlind(SUNAdaptController C, sunrealtype bias)
+{
+  /* set allowed value, otherwise set default */
+  if (bias <= RCONST(0.0)) {
+    SODERLIND_BIAS(C) = DEFAULT_BIAS;
+  } else {
+    SODERLIND_BIAS(C) = bias;
+  }
+
+  return SUNADAPTCONTROLLER_SUCCESS;
+}
+
+int SUNAdaptController_UpdateH_Soderlind(SUNAdaptController C, sunrealtype h, sunrealtype dsm)
+{
+  SODERLIND_EPP(C) = SODERLIND_EP(C);
+  SODERLIND_EP(C)  = SODERLIND_BIAS(C) * dsm;
+  SODERLIND_HPP(C) = SODERLIND_HP(C);
+  SODERLIND_HP(C)  = h;
+  if (SODERLIND_FIRSTSTEPS(C) < 2) {
+    SODERLIND_FIRSTSTEPS(C) += 1;
+  }
+  return SUNADAPTCONTROLLER_SUCCESS;
+}
+
+int SUNAdaptController_Space_Soderlind(SUNAdaptController C, long int* lenrw, long int* leniw)
+{
```

Comment:
```suggestion
{
  if (C == NULL || lenrw == NULL || leniw == NULL) { return SUNADAPTCONTROLLER_ILL_INPUT; }
```

Reply:
It is impossible for `C` to equal `NULL` here, since if it were then the base class would never have called this routine, and the checks on `lenrw` and `leniw` are redundant since you added those to the base class.

Reply:
Although it's not the typical use case, this function can be called directly so I think we should still check the inputs.